### PR TITLE
drivers: bbram: Add bbram driver for mec device

### PR DESCRIPTION
**Pull request by jvasanth1**
Add bbram driver for Microchip mec device

Signed-off-by: Jay Vasanth <jay.vasanth@microchip.com>

### CODE REVIEW

File: drivers/bbram/Kconfig.xec [hunk 1]
```diff
@@ -0,0 +1,12 @@
+# Copyright (c) 2021 Google Inc
+# SPDX-License-Identifier: Apache-2.0
+
+DT_COMPAT_ST_BBRAM_XEC := microchip,xec-bbram
+
+config BBRAM_XEC
+	bool "Microchip XEC Battery-backed RAM drivers"
+	depends on SOC_FAMILY_MEC
+	default $(dt_compat_enabled,${DT_COMPAT_ST_BBRAM_XEC})
```
Comment (yperess) [inline drivers/bbram/Kconfig.xec:9]: Not sure what the proper Kconfig rules are but through the code we're using () not {}

File: drivers/bbram/Kconfig.xec [hunk 2]
```diff
@@ -0,0 +1,12 @@
+# Copyright (c) 2021 Google Inc
+# SPDX-License-Identifier: Apache-2.0
```
Comment (yperess) [inline drivers/bbram/Kconfig.xec:2]: Copy paste?

File: drivers/bbram/bbram_xec.c [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+static int bbram_xec_check_invalid(const struct device *dev)
+{
+	struct bbram_xec_config const *dcfg =
+		(const struct bbram_xec_config *)dev->config;
+	struct vbatr_regs *const regs = (struct vbatr_regs *)dcfg->vbr_base;
+
+	if (regs->PFRS & BIT(MCHP_VBATR_PFRS_VBAT_RST_POS)) {
+		regs->PFRS |= BIT(MCHP_VBATR_PFRS_VBAT_RST_POS);
+		return 1; /* VBAT power rail failure */
```
Comment (yperess) [inline drivers/bbram/bbram_xec.c:37]: Errors should be negative and use errno.h. Will `-EFAULT` work here?

**Review by yperess — CHANGES_REQUESTED** [review]

**Review by jackrosenthal — DISMISSED** [review]

File: drivers/bbram/bbram_xec.c [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+	uintptr_t base;
+	/** BBRAM size (Unit:bytes) */
+	int size;
+};
+
+static int bbram_xec_check_invalid(const struct device *dev)
+{
+	struct bbram_xec_config const *dcfg =
+		(const struct bbram_xec_config *)dev->config;
+	struct vbatr_regs *const regs = (struct vbatr_regs *)dcfg->vbr_base;
```
Comment (gmarull) [inline drivers/bbram/bbram_xec.c:33]: why is `const` qualifier dropped?

why not store a pointer to `struct vbatr_regs *`?

File: drivers/bbram/bbram_xec.c [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+#include <drivers/bbram.h>
+#include <errno.h>
+#include <soc.h>
+#include <sys/util.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(bbram, CONFIG_BBRAM_LOG_LEVEL);
+
+#define XEC_VBAT_REGS_ADDR \
+	(DT_REG_ADDR_BY_NAME(DT_INST(0, microchip_xec_pcr), vbatr))
```
Comment (gmarull) [inline drivers/bbram/bbram_xec.c:18]: driver seems multi-instance, why `0` here?

**Review by gmarull — CHANGES_REQUESTED** [review]

File: drivers/bbram/bbram_xec.c (hunk 3 above)
Comment (jvasanth1) [reply drivers/bbram/bbram_xec.c:37]: yes, will update to -EFAULT

File: drivers/bbram/bbram_xec.c (hunk 5 above)
Comment (jvasanth1) [reply drivers/bbram/bbram_xec.c:18]: The vbr_base is not instance specific to the bbram, and hence we will remove it from the bbram_xec_config structure and in bbram_xec_check_invalid(), we will refer to the registers as

`struct vbatr_regs *const regs = (const struct vbatr_regs *)(DT_REG_ADDR_BY_NAME(DT_INST(0, microchip_xec_pcr), vbatr))`

Comment (jvasanth1) [reply drivers/bbram/bbram_xec.c:18]: Please confirm and I can update accordingly.

Comment (gmarull) [reply drivers/bbram/bbram_xec.c:18]: If some registers are common for all devices, then it makes sense, yes.

File: drivers/bbram/Kconfig.xec (hunk 1 above)
Comment (jvasanth1) [reply drivers/bbram/Kconfig.xec]: Fixed in latest commit

File: drivers/bbram/bbram_xec.c (hunk 3 above)
Comment (jvasanth1) [reply drivers/bbram/bbram_xec.c]: Fixed in latest commit

File: drivers/bbram/bbram_xec.c (hunk 4 above)
Comment (jvasanth1) [reply drivers/bbram/bbram_xec.c]: Fixed in latest commit.

File: drivers/bbram/bbram_xec.c (hunk 5 above)
Comment (jvasanth1) [reply drivers/bbram/bbram_xec.c]: Fixed in latest commit.

File: drivers/bbram/Kconfig.xec (hunk 2 above)
Comment (jvasanth1) [reply drivers/bbram/Kconfig.xec]: Fixed.

**Comment by jvasanth1** [comment]
Fixed all the review comments. Please check and approve.

File: drivers/bbram/bbram_xec.c [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+		return -EFAULT; /* VBAT power rail failure */
+	}
+
+	return 0;
+}
+
+static int bbram_xec_get_size(const struct device *dev, size_t *size)
+{
+	const struct bbram_xec_config *dcfg =
+		(const struct bbram_xec_config *)dev->config;
```
Comment (gmarull) [inline drivers/bbram/bbram_xec.c:41]: please remove redundant casts:

```suggestion
		dev->config;
```

File: drivers/bbram/bbram_xec.c [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+			  uint8_t *data)
+{
+	const struct bbram_xec_config *dcfg =
+		(const struct bbram_xec_config *)dev->config;
+
+	if (size < 1 || offset + size > dcfg->size) {
+		return -EFAULT;
+	}
+
+	bytecpy(data, ((uint8_t *)dcfg->base + offset), size);
```
Comment (gmarull) [inline drivers/bbram/bbram_xec.c:57]: why not define `base` using the right type? (seems `uint8_t *`).

File: drivers/bbram/bbram_xec.c [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+#define DT_DRV_COMPAT microchip_xec_bbram
+
+#include <drivers/bbram.h>
+#include <errno.h>
+#include <soc.h>
+#include <sys/util.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(bbram, CONFIG_BBRAM_LOG_LEVEL);
```
Comment (gmarull) [inline drivers/bbram/bbram_xec.c:15]: this module doesn't seem to log anything

File: drivers/bbram/bbram_xec.c (hunk 6 above)
Comment (jvasanth1) [reply drivers/bbram/bbram_xec.c]: removed redundant cast

File: drivers/bbram/bbram_xec.c (hunk 7 above)
Comment (jvasanth1) [reply drivers/bbram/bbram_xec.c]: defined base as `uint8_t *`

**Comment by jvasanth1** [comment]
Fixed latest review comments. Please check and approve.

**Review by gmarull — APPROVED** [review]
Thanks, LGTM

**Comment by jvasanth1** [comment]
@yperess Please review and approve.

**Review by yperess — APPROVED** [review]

**Review by tejlmand — APPROVED** [review]

**Pull request merged**